crates/proof-of-sql/src/evm_compatibility/serialize_query_expr.rs

```diff
@@ -0,0 +1,158 @@
+use super::dyn_proof_plan_serializer::{DynProofPlanSerializer, ProofPlanSerializationError};
+use crate::{
+    base::scalar::Scalar,
+    sql::{parse::QueryExpr, proof::ProofPlan},
+};
+use alloc::vec::Vec;
+
+/// Serializes a `QueryExpr` into a vector of bytes.
+///
+/// This function takes a `QueryExpr` and attempts to serialize it into a vector of bytes.
+/// The serialization is done in a manner that is compatible with efficient proof verification
+/// on the EVM.
+///
+/// # Arguments
+///
+/// * `query_expr` - A reference to the `QueryExpr` to be serialized.
+///
+/// # Returns
+///
+/// * `Ok(Vec<u8>)` - A vector of bytes representing the serialized query expression.
+/// * `Err(ProofPlanSerializationError)` - An error indicating why the serialization failed.
+///
+/// # Errors
+///
+/// This function returns a `ProofPlanSerializationError::NotSupported` error if the query
+/// expression contains postprocessing steps or if the proof plan cannot be serialized.
+pub fn serialize_query_expr<S: Scalar>(
+    query_expr: &QueryExpr,
+) -> Result<Vec<u8>, ProofPlanSerializationError> {
+    let plan = query_expr
+        .postprocessing()
+        .is_empty()
+        .then(|| query_expr.proof_expr())
+        .ok_or(ProofPlanSerializationError::NotSupported)?;
+    let bytes = DynProofPlanSerializer::<S>::try_new(
+        plan.get_table_references(),
+        plan.get_column_references(),
+    )?
+    .serialize_dyn_proof_plan(plan)?
+    .into_bytes();
+    Ok(bytes)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        base::{
+            database::{ColumnRef, ColumnType, LiteralValue},
+            map::indexset,
+            scalar::test_scalar::TestScalar,
+        },
+        evm_compatibility::{
+            dyn_proof_plan_serializer::{
+                constants::*, DynProofPlanSerializer, ProofPlanSerializationError,
+            },
+            serialize_query_expr,
+        },
+        sql::{
+            parse::QueryExpr,
+            postprocessing::{OwnedTablePostprocessing, SlicePostprocessing},
+            proof_exprs::{
+                AliasedDynProofExpr, ColumnExpr, DynProofExpr, EqualsExpr, LiteralExpr, TableExpr,
+            },
+            proof_plans::{DynProofPlan, EmptyExec, FilterExec},
+        },
+    };
+    use core::iter;
+    use itertools::Itertools;
+
+    #[test]
+    fn we_can_generate_serialized_proof_plan_for_query_expr() {
+        let table_ref = "namespace.table".parse().unwrap();
+        let identifier_alias = "alias".parse().unwrap();
```

Comment:
what is `identifier_alias` used for?

Reply:
Nothing at the moment. Eventually, it will need to be used to validate the result names, if desired. But, it isn't going to be used by the EVM proof verifier.